### PR TITLE
Fix failing enforce in `constantResolutionFailed`

### LINKED ISSUE #1126: (From Fuzzer) resolver/resolver.cc:234 enforced condition !resolved.exists() has failed: (no message provided)
#### Input

[→ View on sorbet.run](https://sorbet.run/#true%0A%20%20%0AA%20%3D%20T.type_alias(y(S))%0AC%20%3D%20A%3A%3AB%0A%23%20%20%20st)

```ruby
true
  
A = T.type_alias(y(S))
C = A::B
#   st
```

#### Observed output

```
Exception::raise(): resolver/resolver.cc:234 enforced condition !resolved.exists() has failed: (no message provided)

Backtrace:
sorbet::Exception::enforce_handler(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) (in librealmain.so) (Exception.h:47)
sorbet::resolver::(anonymous namespace)::ResolveConstantsWalk::constantResolutionFailed(sorbet::core::MutableContext, sorbet::resolver::(anonymous namespace)::ResolveConstantsWalk::ResolutionItem&) (in libresolver.so) (resolver.cc:234)
sorbet::resolver::(anonymous namespace)::ResolveConstantsWalk::resolveConstants(sorbet::core::MutableContext, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::WorkerPool&) (in libresolver.so) (resolver.cc:774)
sorbet::resolver::Resolver::run(sorbet::core::MutableContext, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::WorkerPool&) (in libresolver.so) (resolver.cc:1554)
sorbet::realmain::pipeline::resolve(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&, bool) (in libpipeline.so) (pipeline.cc:859)
sorbet::realmain::realmain(int, char**) (in librealmain.so) (realmain.cc:429)
main (in sorbet) (main.cc:7)
start (in libdyld.dylib) + 1
0x00000002 (in sorbet)

min_crashers/crash-163a2ebdf77f22cb81e406be137a4775d9ce661f.done:3: Unable to resolve right hand side of type alias A https://srb.help/5024
     3 |A = T.type_alias(y(S))
        ^
    min_crashers/crash-163a2ebdf77f22cb81e406be137a4775d9ce661f.done:3: Type alias used here
     3 |A = T.type_alias(y(S))
        ^

min_crashers/crash-163a2ebdf77f22cb81e406be137a4775d9ce661f.done:3: Unable to resolve constant S https://srb.help/5002
     3 |A = T.type_alias(y(S))
                           ^
    min_crashers/crash-163a2ebdf77f22cb81e406be137a4775d9ce661f.done:4: Did you mean: C?
     4 |C = A::B
        ^
    min_crashers/crash-163a2ebdf77f22cb81e406be137a4775d9ce661f.done:3: Did you mean: A?
     3 |A = T.type_alias(y(S))
        ^
    https://github.com/sorbet/sorbet/tree/master/rbi/sorbet/t.rbi#L16: Did you mean: T?
    16 |module T
        ^^^^^^^^

min_crashers/crash-163a2ebdf77f22cb81e406be137a4775d9ce661f.done:4: Resolving constants through type aliases is not supported https://srb.help/5031
     4 |C = A::B
            ^^^^

???: Exception resolving (backtrace is above) https://srb.help/1001
Errors: 4
```

#### Expected behavior

Exception does not occur.

### PR DESCRIPTION
**Pull request by aisamanra**
### Motivation
Fixes #1126: the previous code assumed that failed constant resolutions that still produced symbols that existed must be because of bad type aliases, but if you tried writing a constant alias that in turn referenced an ill-formed type alias (e.g. `A = T.type_alias(x); B = A::X`) then we'd run `constantResolutionFailed` on an existing class symbol and run into a failed enforce. As far as I can tell, this only happens when several other errors have already been raised, and those produce useful error messages in this case as well, so this just adds stub code to handle this esoteric case.

### Test plan

See included automated tests. Included cleaned-up fuzzed test.

### CODE REVIEW

File: resolver/resolver.cc [hunk 1]
```diff
@@ -230,6 +230,13 @@ class ResolveConstantsWalk {
             }
             job.out->symbol = resolved;
             return;
+        } else if (resolved.exists() && resolved.data(ctx)->isClassClass()) {
+            // this can happen when we try to resolve a constant through a type alias which was itself not
+            // well-formed. If we've gotten here, we already have shown reasonable error messages to the user, so just
+            // make sure we can continue gracefully by stubbing in an untyped
+            resolved.data(ctx)->resultType = core::Types::untyped(ctx, resolved);
```
Comment (DarkDimius) [inline resolver/resolver.cc]: I'm concerned of this branch being hit in other surprising cases.
Can we emit another error here?

Comment (DarkDimius) [reply resolver/resolver.cc]: I also am not sure if thats the right way to handle this. 
The fact that handling the previous error has created a case that wasn't handled by other code suggests to me that the buggy code is the code that recovers from error.
Could you see if we can change that code instead?

**Comment by aisamanra** [comment]
The final fix is a bit different. For context, the failing program here was
```ruby
A = T.type_alias(A)
C = A::B
```
The problem with the code before was that we would be unable to resolve the constants `A` (which was known to be a type alias, but not any specific one) or `A::B`, so our class alias pass would make no progress (because the RHS of `C = A::B` was unresolved) and our type alias pass would make no progress (because the rightmost `A` in `A = T.type_alias(A)` was unresolved—of course, because we were trying to resolve it.) However, because we already knew that `A` was a type alias—we just didn't know what it as an alias for yet—we should have at least been able to discover that `A::B` was an invalid constant because you can't access constants under type aliases, but while looking at the scope `A::` we'd run into this branch https://github.com/sorbet/sorbet/blob/8014895b2562273fee34dc81d468cd30817b9ff7/resolver/resolver.cc#L202-L205
because the scope wasn't yet associated with the (unresolved but still present) symbol `A`. Proactively resolving the scope and fixing that TODO meant that we would realize at least that `A::B` was not valid, which in turn meant we could "resolve" `C = A::B` as being invalid earlier and prevent us from getting into the bad state we were getting in before.

File: resolver/resolver.cc [hunk 2]
```diff
@@ -192,17 +192,19 @@ class ResolveConstantsWalk {
...
+                }
+            }
             auto sym = id->symbol;
             if (sym.exists() && sym.data(ctx)->isTypeAlias()) {
                 if (auto e = ctx.state.beginError(c->loc, core::errors::Resolver::ConstantInTypeAlias)) {
                     e.setHeader("Resolving constants through type aliases is not supported");
                 }
                 return core::Symbols::untyped();
             }
-            if (!id->symbol.exists()) {
```
Comment (DarkDimius) [inline resolver/resolver.cc]: why was this removed?

File: resolver/resolver.cc [hunk 3]
```diff
@@ -192,17 +192,19 @@ class ResolveConstantsWalk {
         }
         ast::Expression *resolvedScope = c->scope.get();
         if (auto *id = ast::cast_tree<ast::ConstantLit>(resolvedScope)) {
+            if (!id->symbol.exists()) {
+                id->symbol = resolveConstant(ctx, nesting, id->original);
```
Comment (DarkDimius) [inline resolver/resolver.cc]: So far we've been intentionally trying to make `resolveConstant` non recursive.
Could we keep it lazy?

File: resolver/resolver.cc (hunk 2 above)
Comment (aisamanra) [reply resolver/resolver.cc]: This check just got moved up earlier.

**Comment by aisamanra** [comment]
This now has a slightly different implementation that's not recursive. I have a mild bit of FUD about it, though, and have an idea to address that FUD but want to check first—this introduces to the constant resolver that we might be "making progress" (i.e. making updates to a constant resolution job) without having finished that job, but that's still kind of implicit (i.e. we still only repeat the constant loop if we've actually resolved jobs.) To address this, we could introduce an _explicit_ notion of "we're progressing but not removing any resolved jobs" when we update information in a job. At the same time, this complicates the implementation of an already-complicated part of the code and I'm not 100% sure we need it yet.

**Comment by DarkDimius** [comment]
IRL'd on it, agreed on a way to address this that seems to not cause that FUD

**Pull request merged**